lib/Dialect/Comb/CombFolds.cpp

```diff
@@ -73,6 +74,113 @@ static bool tryFlatteningOperands(Op op, PatternRewriter &rewriter) {
   return false;
 }
 
+// Given a range of uses of an operation, find the lowest and highest bits
+// inclusive that are ever referenced. The range of uses must not be empty.
+static std::pair<size_t, size_t>
+getLowestBitAndHighestBitRequired(Operation::use_range uses, bool narrowTrailingBits, size_t originalOpWidth)
+{
+  assert(!uses.empty() &&
+      "getLowestBitAndHighestBitRequired cannot operate on a empty list of uses.");
+
+  // when we don't want to narrowTrailingBits (namely in arithmetic
+  // operations), forcing lowestBitRequired = 0
+  size_t lowestBitRequired = narrowTrailingBits ? originalOpWidth - 1 : 0;
+  size_t highestBitRequired = 0;
+
+  for (auto &use : uses) {
+    if (auto extractOp = dyn_cast<ExtractOp>(use.getOwner())) {
+      size_t lowBit = extractOp.lowBit();
+      size_t highBit = extractOp.getType().getWidth() + lowBit - 1;
+      highestBitRequired = std::max(highestBitRequired, highBit);
+      lowestBitRequired = std::min(lowestBitRequired, lowBit);
+      continue;
+    }
+
+    highestBitRequired = originalOpWidth - 1;
+    lowestBitRequired = 0;
+    break;
+  }
+
+  return {lowestBitRequired, highestBitRequired};
+}
+
+// narrowOperationWidth(...) attempts to create a new instance of Op
+// with createOp using narrowed operands. Ie: transforming f(x, y) into
+// f(x[n:m], y[n:m]), It analyzes all usage sites of narrowingCandidate and
+// mutate them to the newly created narrowed version if it's benefitial.
+//
+// narrowTrailingBits determines whether undemanded trailing bits should
+// be aggressively stripped of.
+//
+// Returns true if IR is mutated. IR is mutated iff narrowing happens.
+template <class Op, class F>
+static bool narrowOperationWidth(Op narrowingCandidate, ValueRange inputs,
+                                 bool narrowTrailingBits,
+                                 PatternRewriter &rewriter, F createOp) {
+  // If the result is never used, no point optimizing this. It will
+  // also complicated error handling in getLowestBitAndHigestBitRequired.
+  Operation::use_range uses = narrowingCandidate.getOperation()->getUses();
+  if (uses.empty()) {
+    return false;
+  }
+
+  size_t highestBitRequired;
+  size_t lowestBitRequired;
+  size_t originalOpWidth = narrowingCandidate.getType().getIntOrFloatBitWidth();
+
+  std::tie(lowestBitRequired, highestBitRequired) =
+    getLowestBitAndHighestBitRequired(uses, narrowTrailingBits, originalOpWidth);
+
+  // Give up, because we can't make it narrower than it already is.
+  if (lowestBitRequired == 0 && highestBitRequired == originalOpWidth - 1) {
+    return false;
+  }
```

Comment:
Minor request, but LLVM style is to omit braces on single statement "if" statements like this.